src/hotspot/share/opto/escape.hpp

```diff
@@ -569,6 +575,23 @@ class ConnectionGraph: public ResourceObj {
   // Compute the escape information
   bool compute_escape();
 
+  void set_not_scalar_replaceable(PointsToNode* ptn DEBUG_ONLY(COMMA const char* reason="reason unknown")) const {
+#ifndef PRODUCT
+    if (_compile->trace_escape_analysis()) {
+        assert(ptn != nullptr, "should not be null");
+        ptn->dump_header(true);
+        tty->print_cr("is NSR. %s", reason);
+    }
+#endif
+    ptn->set_scalar_replaceable(false);
+  }
+
+#ifndef PRODUCT
```

Comment:
Should be `#ifdef ASSERT` since it is called only in `DEBUG_ONLY`.
Or may be better change `DEBUG_ONLY()` macros in call sites and arguments to `NOT_PRODUCT()`. That way we can use it in optimized VM.

Reply:
Thanks. I was wondering if `DEBUG_ONLY` was the right one.

The flag itself in compilerDirectives.hpp uses `NOT_PRODUCT`, so I will switch to that.